src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -3157,6 +3158,49 @@ protected byte[] invert(byte[] bytes, int start,
             return modified;
         }
 
+        @Specialization
+        protected byte[] invert(InternalByteArray byteArray, int start,
+                @Cached BranchProfile foundLowerCaseCharProfile,
+                @Cached BranchProfile foundUpperCaseCharProfile,
+                @Cached LoopConditionProfile loopProfile) {
+            var bytes = byteArray.getArray();
+            byte[] modified = null;
+
+            int i = start + byteArray.getOffset();
+            try {
+                for (; loopProfile.inject(i < byteArray.getEnd()); i++) {
+                    final byte b = bytes[i];
+
+                    if (lowerToUpper && StringSupport.isAsciiLowercase(b)) {
+                        foundLowerCaseCharProfile.enter();
+
+                        if (modified == null) {
+                            modified = bytes.clone();
```

Comment:
This feels suboptimal, because it will also copy bytes before getOffset() or after getEnd().
I think CopyToByteArrayNode would be better here, and easier to use (the returned `byte[]` is all relevant bytes, no extra bytes).

Reply:
This is a case where I kept what was there already. I can change, but I'm trying to keep the scope down. It looks new, but it's just a copy-and-paste of the earlier `invert` call that didn't take an `InternalByteArray`.

Reply:
I see it as related to adopting InternalByteArray and its offset/length/end, so I think we should do this as part of the migration.
There are hopefully few other `.clone()` calls.
We could do it later though, in that case then please leave a `// TODO: avoid copying before offset/after end` there.

Reply:
I'll look at it, but it's a bit messier than it looks at first glance. The idea is to not copy the array unless it's necessary. Using `CopyToByteArrayNode` will copy just the range we need, effectively setting the offset back to `0`, which affects the loop index. Using `clone`, we don't have to adjust the loop bounds.

Reply:
Yes, I think the cleanest is `i` does not include `byteArray.getOffset()` and rather that offset is only added when reading from `byteArray.getArray()`.
We could imagine a worst-case of a 1MB String, and then a substring of 10 bytes is downcased, there the `.clone()` is quite expensive + it leaks memory forever and in a hidden way (held by the TruffleString of FromByteArrayNode).
There is already a leak due to lazy substrings, but this adds a second unexpected one (if we mutate at least we shouldn't use more bytes than needed to represent the result).
Such a leak seems a big problem, hence I think we need to solve this.

An alternative would be to just eagerly do the `FromByteArrayNode` (that avoids any leak), but it's of course a bit suboptimal if there is nothing to change (doesn't change time complexity, still O(n)). Might even be better as then the loop is less branchy. OTOH it would potentially increase memory usage if nothing to do and there is both the original TruffleString and this substring alive, so not ideal.

Reply:
Done in 050be7813fe6194100b81668cd991ecc1eb64b20.